Test hanging with cbl-views.js functional test

Using 1.3.0-35 (ForestDB + Encryption), API 23 x86

cbl-views.js test hang. I'm seeing the below output. Perhaps this is a red herring but wanted to file regardless. 

```
07-18 19:23:28.648  5549  5570 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.648  5549  5570 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.648  5549  5570 V Listener: CountDownLatch.await() START
07-18 19:23:28.648  5549  5570 V Listener: CountDownLatch.await() END
07-18 19:23:28.655  5549  5568 D Listener: authHeader is null
07-18 19:23:28.655  5549  5568 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.655  5549  5568 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.655  5549  5568 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.655  5549  5568 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.655  5549  5568 V Listener: CountDownLatch.await() START
07-18 19:23:28.655  5549  5568 V Listener: CountDownLatch.await() END
07-18 19:23:28.662  5549  5567 D Listener: authHeader is null
07-18 19:23:28.662  5549  5567 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.662  5549  5567 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.662  5549  5567 V Database: DELETE _id=cbl_views_5, _rev=1-aaa6c063924b64a141c98820efcc0022 (allowConflict=false)
07-18 19:23:28.662  5549  5567 I cbforest: Database: beginTransaction
07-18 19:23:28.662  5549  5567 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f35 --> 0000001c0001831102588567fcabdfacc722da0a6e0d3df013007b7d0000001dffff401101aaa6c063924b64a141c98820efcc00220680a00700000000 (meta 011102588567fcabdfacc722da0a6e0d3df01300) (seq 17)
07-18 19:23:28.662  5549  5567 I cbforest: Database: commit transaction
07-18 19:23:28.667  5549  5567 V CBLite  : ---> Added: {cbl_views_5 #2-588567fcabdfacc722da0a6e0d3df013 @17 DEL} as seq 17
07-18 19:23:28.667  5549  5567 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.667  5549  5567 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.667  5549  5567 V Listener: CountDownLatch.await() START
07-18 19:23:28.667  5549  5567 V Listener: CountDownLatch.await() END
07-18 19:23:28.674  5549  5569 D Listener: authHeader is null
07-18 19:23:28.674  5549  5569 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.674  5549  5569 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.674  5549  5569 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.674  5549  5569 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.674  5549  5569 V Listener: CountDownLatch.await() START
07-18 19:23:28.674  5549  5569 V Listener: CountDownLatch.await() END
07-18 19:23:28.680  5549  5571 D Listener: authHeader is null
07-18 19:23:28.680  5549  5571 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.685  5549  5571 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.685  5549  5571 V Database: DELETE _id=cbl_views_6, _rev=1-aaa6c063924b64a141c98820efcc0022 (allowConflict=false)
07-18 19:23:28.685  5549  5571 I cbforest: Database: beginTransaction
07-18 19:23:28.686  5549  5571 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f36 --> 0000001c0001831102588567fcabdfacc722da0a6e0d3df013007b7d0000001dffff401101aaa6c063924b64a141c98820efcc00220780e00700000000 (meta 011102588567fcabdfacc722da0a6e0d3df01300) (seq 18)
07-18 19:23:28.686  5549  5571 I cbforest: Database: commit transaction
07-18 19:23:28.689  5549  5571 V CBLite  : ---> Added: {cbl_views_6 #2-588567fcabdfacc722da0a6e0d3df013 @18 DEL} as seq 18
07-18 19:23:28.689  5549  5571 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.689  5549  5571 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.689  5549  5571 V Listener: CountDownLatch.await() START
07-18 19:23:28.689  5549  5571 V Listener: CountDownLatch.await() END
07-18 19:23:28.707  5549  5570 D Listener: authHeader is null
07-18 19:23:28.707  5549  5570 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.708  5549  5570 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.708  5549  5570 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.708  5549  5570 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.708  5549  5570 V Listener: CountDownLatch.await() START
07-18 19:23:28.708  5549  5570 V Listener: CountDownLatch.await() END
07-18 19:23:28.719  5549  5568 D Listener: authHeader is null
07-18 19:23:28.719  5549  5568 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.719  5549  5568 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.720  5549  5568 V Database: DELETE _id=cbl_views_7, _rev=1-aaa6c063924b64a141c98820efcc0022 (allowConflict=false)
07-18 19:23:28.720  5549  5568 I cbforest: Database: beginTransaction
07-18 19:23:28.720  5549  5568 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f37 --> 0000001c0001831102588567fcabdfacc722da0a6e0d3df013007b7d0000001dffff401101aaa6c063924b64a141c98820efcc00220880a00800000000 (meta 011102588567fcabdfacc722da0a6e0d3df01300) (seq 19)
07-18 19:23:28.720  5549  5568 I cbforest: Database: commit transaction
07-18 19:23:28.724  5549  5568 V CBLite  : ---> Added: {cbl_views_7 #2-588567fcabdfacc722da0a6e0d3df013 @19 DEL} as seq 19
07-18 19:23:28.724  5549  5568 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.724  5549  5568 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.724  5549  5568 V Listener: CountDownLatch.await() START
07-18 19:23:28.724  5549  5568 V Listener: CountDownLatch.await() END
07-18 19:23:28.737  5549  5567 D Listener: authHeader is null
07-18 19:23:28.737  5549  5567 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.737  5549  5567 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.737  5549  5567 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.737  5549  5567 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.737  5549  5567 V Listener: CountDownLatch.await() START
07-18 19:23:28.737  5549  5567 V Listener: CountDownLatch.await() END
07-18 19:23:28.747  5549  5569 D Listener: authHeader is null
07-18 19:23:28.747  5549  5569 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.747  5549  5569 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.748  5549  5569 V Database: DELETE _id=cbl_views_8, _rev=1-aaa6c063924b64a141c98820efcc0022 (allowConflict=false)
07-18 19:23:28.748  5549  5569 I cbforest: Database: beginTransaction
07-18 19:23:28.748  5549  5569 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f38 --> 0000001c0001831102588567fcabdfacc722da0a6e0d3df013007b7d0000001dffff401101aaa6c063924b64a141c98820efcc00220980e00800000000 (meta 011102588567fcabdfacc722da0a6e0d3df01300) (seq 20)
07-18 19:23:28.748  5549  5569 I cbforest: Database: commit transaction
07-18 19:23:28.755  5549  5569 V CBLite  : ---> Added: {cbl_views_8 #2-588567fcabdfacc722da0a6e0d3df013 @20 DEL} as seq 20
07-18 19:23:28.756  5549  5569 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.756  5549  5569 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.756  5549  5569 V Listener: CountDownLatch.await() START
07-18 19:23:28.756  5549  5569 V Listener: CountDownLatch.await() END
07-18 19:23:28.767  5549  5571 D Listener: authHeader is null
07-18 19:23:28.767  5549  5571 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.767  5549  5571 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.767  5549  5571 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.767  5549  5571 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.767  5549  5571 V Listener: CountDownLatch.await() START
07-18 19:23:28.767  5549  5571 V Listener: CountDownLatch.await() END
07-18 19:23:28.776  5549  5570 D Listener: authHeader is null
07-18 19:23:28.776  5549  5570 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.776  5549  5570 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.776  5549  5570 V Database: DELETE _id=cbl_views_9, _rev=1-aaa6c063924b64a141c98820efcc0022 (allowConflict=false)
07-18 19:23:28.776  5549  5570 I cbforest: Database: beginTransaction
07-18 19:23:28.776  5549  5570 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f39 --> 0000001c0001831102588567fcabdfacc722da0a6e0d3df013007b7d0000001dffff401101aaa6c063924b64a141c98820efcc00220a80a00900000000 (meta 011102588567fcabdfacc722da0a6e0d3df01300) (seq 21)
07-18 19:23:28.776  5549  5570 I cbforest: Database: commit transaction
07-18 19:23:28.783  5549  5570 V CBLite  : ---> Added: {cbl_views_9 #2-588567fcabdfacc722da0a6e0d3df013 @21 DEL} as seq 21
07-18 19:23:28.783  5549  5570 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.783  5549  5570 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.783  5549  5570 V Listener: CountDownLatch.await() START
07-18 19:23:28.783  5549  5570 V Listener: CountDownLatch.await() END
07-18 19:23:28.798  5549  5568 D Listener: authHeader is null
07-18 19:23:28.801  5549  5568 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.802  5549  5568 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.812  5549  5568 V Database: PUT _id=cbl_views_0, _rev=null (allowConflict=false)
07-18 19:23:28.813  5549  5568 I cbforest: Database: beginTransaction
07-18 19:23:28.813  5549  5568 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f30 --> 0000014200028a1103bb60e13962a953986cc34ad0eec3171b007b226a6f696e6564223a5b323031332c372c305d2c225f6174746163686d656e7473223a7b2270726f66696c652e747874223a7b22646967657374223a22736861312d527978506635646a50344a41306e372f44444c67334e58713970303d222c226c656e677468223a33302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22746578742f706c61696e227d2c2270686f746f2e706e67223a7b22646967657374223a22736861312d64756f5279623454773277504c6242796f513177586d5934572b413d222c226c656e677468223a333435302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22696d6167652f706e67227d7d2c22706f696e7473223a307d0000001dffff401101aaa6c063924b64a141c98820efcc00220180e0040000001d0001411102588567fcabdfacc722da0a6e0d3df0130c80a00a00000000 (meta 041103bb60e13962a953986cc34ad0eec3171b00) (seq 22)
07-18 19:23:28.813  5549  5568 I cbforest: Database: commit transaction
07-18 19:23:28.817  5549  5568 V CBLite  : ---> Added: {cbl_views_0 #3-bb60e13962a953986cc34ad0eec3171b @22} as seq 22
07-18 19:23:28.817  5549  5568 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.817  5549  5568 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.817  5549  5568 V Listener: CountDownLatch.await() START
07-18 19:23:28.817  5549  5568 V Listener: CountDownLatch.await() END
07-18 19:23:28.833  5549  5567 D Listener: authHeader is null
07-18 19:23:28.833  5549  5567 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.833  5549  5567 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.834  5549  5567 V Database: PUT _id=cbl_views_1, _rev=null (allowConflict=false)
07-18 19:23:28.834  5549  5567 I cbforest: Database: beginTransaction
07-18 19:23:28.834  5549  5567 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f31 --> 0000014200028a1103826ea164e69e6433d705db61424ae1e6007b226a6f696e6564223a5b323031332c372c315d2c225f6174746163686d656e7473223a7b2270726f66696c652e747874223a7b22646967657374223a22736861312d527978506635646a50344a41306e372f44444c67334e58713970303d222c226c656e677468223a33302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22746578742f706c61696e227d2c2270686f746f2e706e67223a7b22646967657374223a22736861312d64756f5279623454773277504c6242796f513177586d5934572b413d222c226c656e677468223a333435302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22696d6167652f706e67227d7d2c22706f696e7473223a317d0000001dffff401101aaa6c063924b64a141c98820efcc00220280a0050000001d0001411102588567fcabdfacc722da0a6e0d3df0130d80e00a00000000 (meta 041103826ea164e69e6433d705db61424ae1e600) (seq 23)
07-18 19:23:28.834  5549  5567 I cbforest: Database: commit transaction
07-18 19:23:28.838  5549  5567 V CBLite  : ---> Added: {cbl_views_1 #3-826ea164e69e6433d705db61424ae1e6 @23} as seq 23
07-18 19:23:28.838  5549  5567 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.838  5549  5567 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.838  5549  5567 V Listener: CountDownLatch.await() START
07-18 19:23:28.838  5549  5567 V Listener: CountDownLatch.await() END
07-18 19:23:28.852  5549  5569 D Listener: authHeader is null
07-18 19:23:28.852  5549  5569 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.852  5549  5569 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.852  5549  5569 V Database: PUT _id=cbl_views_2, _rev=null (allowConflict=false)
07-18 19:23:28.852  5549  5569 I cbforest: Database: beginTransaction
07-18 19:23:28.853  5549  5569 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f32 --> 0000014200028a1103f02f8135d2d6587cd6568e27bb338752007b226a6f696e6564223a5b323031332c372c325d2c225f6174746163686d656e7473223a7b2270726f66696c652e747874223a7b22646967657374223a22736861312d527978506635646a50344a41306e372f44444c67334e58713970303d222c226c656e677468223a33302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22746578742f706c61696e227d2c2270686f746f2e706e67223a7b22646967657374223a22736861312d64756f5279623454773277504c6242796f513177586d5934572b413d222c226c656e677468223a333435302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22696d6167652f706e67227d7d2c22706f696e7473223a327d0000001dffff401101aaa6c063924b64a141c98820efcc00220380e0050000001d0001411102588567fcabdfacc722da0a6e0d3df0130e80a00b00000000 (meta 041103f02f8135d2d6587cd6568e27bb33875200) (seq 24)
07-18 19:23:28.853  5549  5569 I cbforest: Database: commit transaction
07-18 19:23:28.855  5549  5569 V CBLite  : ---> Added: {cbl_views_2 #3-f02f8135d2d6587cd6568e27bb338752 @24} as seq 24
07-18 19:23:28.855  5549  5569 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.856  5549  5569 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.856  5549  5569 V Listener: CountDownLatch.await() START
07-18 19:23:28.856  5549  5569 V Listener: CountDownLatch.await() END
07-18 19:23:28.865  5549  5571 D Listener: authHeader is null
07-18 19:23:28.865  5549  5571 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.865  5549  5571 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.865  5549  5571 V Database: PUT _id=cbl_views_3, _rev=null (allowConflict=false)
07-18 19:23:28.865  5549  5571 I cbforest: Database: beginTransaction
07-18 19:23:28.866  5549  5571 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f33 --> 0000014200028a11031be1d747c5f2cba8ef25af28130dfc67007b226a6f696e6564223a5b323031332c372c335d2c225f6174746163686d656e7473223a7b2270726f66696c652e747874223a7b22646967657374223a22736861312d527978506635646a50344a41306e372f44444c67334e58713970303d222c226c656e677468223a33302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22746578742f706c61696e227d2c2270686f746f2e706e67223a7b22646967657374223a22736861312d64756f5279623454773277504c6242796f513177586d5934572b413d222c226c656e677468223a333435302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22696d6167652f706e67227d7d2c22706f696e7473223a337d0000001dffff401101aaa6c063924b64a141c98820efcc00220480a0060000001d0001411102588567fcabdfacc722da0a6e0d3df0130f80e00b00000000 (meta 0411031be1d747c5f2cba8ef25af28130dfc6700) (seq 25)
07-18 19:23:28.866  5549  5571 I cbforest: Database: commit transaction
07-18 19:23:28.870  5549  5571 V CBLite  : ---> Added: {cbl_views_3 #3-1be1d747c5f2cba8ef25af28130dfc67 @25} as seq 25
07-18 19:23:28.870  5549  5571 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.870  5549  5571 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.870  5549  5571 V Listener: CountDownLatch.await() START
07-18 19:23:28.870  5549  5571 V Listener: CountDownLatch.await() END
07-18 19:23:28.880  5549  5570 D Listener: authHeader is null
07-18 19:23:28.880  5549  5570 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.880  5549  5570 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.881  5549  5570 V Database: PUT _id=cbl_views_4, _rev=null (allowConflict=false)
07-18 19:23:28.882  5549  5570 I cbforest: Database: beginTransaction
07-18 19:23:28.882  5549  5570 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f34 --> 0000014200028a11030068e22da1aab2bd7b32cd808918f7c8007b226a6f696e6564223a5b323031332c372c345d2c225f6174746163686d656e7473223a7b2270726f66696c652e747874223a7b22646967657374223a22736861312d527978506635646a50344a41306e372f44444c67334e58713970303d222c226c656e677468223a33302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22746578742f706c61696e227d2c2270686f746f2e706e67223a7b22646967657374223a22736861312d64756f5279623454773277504c6242796f513177586d5934572b413d222c226c656e677468223a333435302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22696d6167652f706e67227d7d2c22706f696e7473223a347d0000001dffff401101aaa6c063924b64a141c98820efcc00220580e0060000001d0001411102588567fcabdfacc722da0a6e0d3df0131080a00c00000000 (meta 0411030068e22da1aab2bd7b32cd808918f7c800) (seq 26)
07-18 19:23:28.882  5549  5570 I cbforest: Database: commit transaction
07-18 19:23:28.886  5549  5570 V CBLite  : ---> Added: {cbl_views_4 #3-0068e22da1aab2bd7b32cd808918f7c8 @26} as seq 26
07-18 19:23:28.886  5549  5570 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.886  5549  5570 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.886  5549  5570 V Listener: CountDownLatch.await() START
07-18 19:23:28.886  5549  5570 V Listener: CountDownLatch.await() END
07-18 19:23:28.901  5549  5568 D Listener: authHeader is null
07-18 19:23:28.901  5549  5568 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.901  5549  5568 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.902  5549  5568 V Database: PUT _id=cbl_views_5, _rev=null (allowConflict=false)
07-18 19:23:28.902  5549  5568 I cbforest: Database: beginTransaction
07-18 19:23:28.902  5549  5568 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f35 --> 0000014200028a1103ebdfff036650b2c6578c19b61668e50d007b226a6f696e6564223a5b323031332c372c355d2c225f6174746163686d656e7473223a7b2270726f66696c652e747874223a7b22646967657374223a22736861312d527978506635646a50344a41306e372f44444c67334e58713970303d222c226c656e677468223a33302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22746578742f706c61696e227d2c2270686f746f2e706e67223a7b22646967657374223a22736861312d64756f5279623454773277504c6242796f513177586d5934572b413d222c226c656e677468223a333435302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22696d6167652f706e67227d7d2c22706f696e7473223a357d0000001dffff401101aaa6c063924b64a141c98820efcc00220680a0070000001d0001411102588567fcabdfacc722da0a6e0d3df0131180e00c00000000 (meta 041103ebdfff036650b2c6578c19b61668e50d00) (seq 27)
07-18 19:23:28.902  5549  5568 I cbforest: Database: commit transaction
07-18 19:23:28.906  5549  5568 V CBLite  : ---> Added: {cbl_views_5 #3-ebdfff036650b2c6578c19b61668e50d @27} as seq 27
07-18 19:23:28.906  5549  5568 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.906  5549  5568 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.906  5549  5568 V Listener: CountDownLatch.await() START
07-18 19:23:28.906  5549  5568 V Listener: CountDownLatch.await() END
07-18 19:23:28.915  5549  5567 D Listener: authHeader is null
07-18 19:23:28.915  5549  5567 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.915  5549  5567 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.916  5549  5567 V Database: PUT _id=cbl_views_6, _rev=null (allowConflict=false)
07-18 19:23:28.916  5549  5567 I cbforest: Database: beginTransaction
07-18 19:23:28.916  5549  5567 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f36 --> 0000014200028a110380e7362bf7ef748966a2cb81f426c71e007b226a6f696e6564223a5b323031332c372c365d2c225f6174746163686d656e7473223a7b2270726f66696c652e747874223a7b22646967657374223a22736861312d527978506635646a50344a41306e372f44444c67334e58713970303d222c226c656e677468223a33302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22746578742f706c61696e227d2c2270686f746f2e706e67223a7b22646967657374223a22736861312d64756f5279623454773277504c6242796f513177586d5934572b413d222c226c656e677468223a333435302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22696d6167652f706e67227d7d2c22706f696e7473223a367d0000001dffff401101aaa6c063924b64a141c98820efcc00220780e0070000001d0001411102588567fcabdfacc722da0a6e0d3df0131280a00d00000000 (meta 04110380e7362bf7ef748966a2cb81f426c71e00) (seq 28)
07-18 19:23:28.916  5549  5567 I cbforest: Database: commit transaction
07-18 19:23:28.920  5549  5567 V CBLite  : ---> Added: {cbl_views_6 #3-80e7362bf7ef748966a2cb81f426c71e @28} as seq 28
07-18 19:23:28.921  5549  5567 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.921  5549  5567 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.921  5549  5567 V Listener: CountDownLatch.await() START
07-18 19:23:28.921  5549  5567 V Listener: CountDownLatch.await() END
07-18 19:23:28.928  5549  5569 D Listener: authHeader is null
07-18 19:23:28.928  5549  5569 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.929  5549  5569 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.929  5549  5569 V Database: PUT _id=cbl_views_7, _rev=null (allowConflict=false)
07-18 19:23:28.929  5549  5569 I cbforest: Database: beginTransaction
07-18 19:23:28.929  5549  5569 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f37 --> 0000014200028a1103b67f633d03dd1514ee1ac5e81976ebeb007b226a6f696e6564223a5b323031332c372c375d2c225f6174746163686d656e7473223a7b2270726f66696c652e747874223a7b22646967657374223a22736861312d527978506635646a50344a41306e372f44444c67334e58713970303d222c226c656e677468223a33302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22746578742f706c61696e227d2c2270686f746f2e706e67223a7b22646967657374223a22736861312d64756f5279623454773277504c6242796f513177586d5934572b413d222c226c656e677468223a333435302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22696d6167652f706e67227d7d2c22706f696e7473223a377d0000001dffff401101aaa6c063924b64a141c98820efcc00220880a0080000001d0001411102588567fcabdfacc722da0a6e0d3df0131380e00d00000000 (meta 041103b67f633d03dd1514ee1ac5e81976ebeb00) (seq 29)
07-18 19:23:28.929  5549  5569 I cbforest: Database: commit transaction
07-18 19:23:28.931  5549  5559 I art     : Background sticky concurrent mark sweep GC freed 4311(778KB) AllocSpace objects, 42(1128KB) LOS objects, 17% free, 3MB/4MB, paused 5.595ms total 46.951ms
07-18 19:23:28.933  5549  5569 V CBLite  : ---> Added: {cbl_views_7 #3-b67f633d03dd1514ee1ac5e81976ebeb @29} as seq 29
07-18 19:23:28.933  5549  5569 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.933  5549  5569 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.933  5549  5569 V Listener: CountDownLatch.await() START
07-18 19:23:28.933  5549  5569 V Listener: CountDownLatch.await() END
07-18 19:23:28.945  5549  5571 D Listener: authHeader is null
07-18 19:23:28.945  5549  5571 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.945  5549  5571 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.945  5549  5571 V Database: PUT _id=cbl_views_8, _rev=null (allowConflict=false)
07-18 19:23:28.946  5549  5571 I cbforest: Database: beginTransaction
07-18 19:23:28.946  5549  5571 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f38 --> 0000014200028a110385984c5a89b570a9f9836c77b649cd20007b226a6f696e6564223a5b323031332c372c385d2c225f6174746163686d656e7473223a7b2270726f66696c652e747874223a7b22646967657374223a22736861312d527978506635646a50344a41306e372f44444c67334e58713970303d222c226c656e677468223a33302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22746578742f706c61696e227d2c2270686f746f2e706e67223a7b22646967657374223a22736861312d64756f5279623454773277504c6242796f513177586d5934572b413d222c226c656e677468223a333435302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22696d6167652f706e67227d7d2c22706f696e7473223a387d0000001dffff401101aaa6c063924b64a141c98820efcc00220980e0080000001d0001411102588567fcabdfacc722da0a6e0d3df0131480a00e00000000 (meta 04110385984c5a89b570a9f9836c77b649cd2000) (seq 30)
07-18 19:23:28.946  5549  5571 I cbforest: Database: commit transaction
07-18 19:23:28.950  5549  5571 V CBLite  : ---> Added: {cbl_views_8 #3-85984c5a89b570a9f9836c77b649cd20 @30} as seq 30
07-18 19:23:28.950  5549  5571 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.950  5549  5571 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.950  5549  5571 V Listener: CountDownLatch.await() START
07-18 19:23:28.950  5549  5571 V Listener: CountDownLatch.await() END
07-18 19:23:28.958  5549  5570 D Listener: authHeader is null
07-18 19:23:28.958  5549  5570 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.958  5549  5570 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.958  5549  5570 V Database: PUT _id=cbl_views_9, _rev=null (allowConflict=false)
07-18 19:23:28.958  5549  5570 I cbforest: Database: beginTransaction
07-18 19:23:28.959  5549  5570 I cbforest: DB 0x7f2fa59c21c0: added 63626c5f76696577735f39 --> 0000014200028a1103fbb17b0586ed05274b23fd0f500b5784007b226a6f696e6564223a5b323031332c372c395d2c225f6174746163686d656e7473223a7b2270726f66696c652e747874223a7b22646967657374223a22736861312d527978506635646a50344a41306e372f44444c67334e58713970303d222c226c656e677468223a33302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22746578742f706c61696e227d2c2270686f746f2e706e67223a7b22646967657374223a22736861312d64756f5279623454773277504c6242796f513177586d5934572b413d222c226c656e677468223a333435302c22726576706f73223a312c2273747562223a747275652c22636f6e74656e745f74797065223a22696d6167652f706e67227d7d2c22706f696e7473223a397d0000001dffff401101aaa6c063924b64a141c98820efcc00220a80a0090000001d0001411102588567fcabdfacc722da0a6e0d3df0131580e00e00000000 (meta 041103fbb17b0586ed05274b23fd0f500b578400) (seq 31)
07-18 19:23:28.959  5549  5570 I cbforest: Database: commit transaction
07-18 19:23:28.962  5549  5570 V CBLite  : ---> Added: {cbl_views_9 #3-fbb17b0586ed05274b23fd0f500b5784 @31} as seq 31
07-18 19:23:28.962  5549  5570 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.962  5549  5570 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.965  5549  5570 V Listener: CountDownLatch.await() START
07-18 19:23:28.965  5549  5570 V Listener: CountDownLatch.await() END
07-18 19:23:28.983  5549  5568 D Listener: authHeader is null
07-18 19:23:28.983  5549  5568 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.983  5549  5568 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.983  5549  5568 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.983  5549  5568 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.983  5549  5568 V Listener: CountDownLatch.await() START
07-18 19:23:28.983  5549  5568 V Listener: CountDownLatch.await() END
07-18 19:23:28.994  5549  5567 D Listener: authHeader is null
07-18 19:23:28.994  5549  5567 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:28.994  5549  5567 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:28.994  5549  5567 V Database: PUT _id=_design/test, _rev=1-1d7f52e75d4a7b3e838c8e677848f3de (allowConflict=false)
07-18 19:23:28.994  5549  5567 I cbforest: Database: beginTransaction
07-18 19:23:28.994  5549  5567 I cbforest: DB 0x7f2fa59c21c0: added 5f64657369676e2f74657374 --> 0000012400018211023488f5d9cf9dfa4496a4234bbbaa6801007b227669657773223a7b226261736963223a7b226d6170223a2266756e6374696f6e28646f6329207b20656d697428646f632e5f69642c20646f632e666f6f29207d227d2c22706c61796572223a7b226d6170223a2266756e6374696f6e28646f6329207b20656d697428646f632e6a6f696e65642c20646f632e706f696e747329207d222c22726564756365223a2266756e6374696f6e286b6579732c2076616c7565732c20726572656475636529207b20726573756c74203d20303b20666f722028693d303b693c76616c7565732e6c656e6774683b692b2b29207b20726573756c74202b3d2076616c7565735b695d207d3b2072657475726e20726573756c7420207d227d7d7d0000001dffff4011011d7f52e75d4a7b3e838c8e677848f3de0b80e00900000000 (meta 0011023488f5d9cf9dfa4496a4234bbbaa680100) (seq 32)
07-18 19:23:28.994  5549  5567 I cbforest: Database: commit transaction
07-18 19:23:28.998  5549  5567 V CBLite  : ---> Added: {_design/test #2-3488f5d9cf9dfa4496a4234bbbaa6801 @32} as seq 32
07-18 19:23:28.998  5549  5567 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:23:28.998  5549  5567 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:23:28.998  5549  5567 V Listener: CountDownLatch.await() START
07-18 19:23:28.998  5549  5567 V Listener: CountDownLatch.await() END
07-18 19:23:29.021  5549  5569 D Listener: authHeader is null
07-18 19:23:29.021  5549  5569 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:29.021  5549  5569 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:29.021  5549  5571 D Listener: authHeader is null
07-18 19:23:29.021  5549  5571 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:29.022  5549  5571 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:29.034  5549  5570 D Listener: authHeader is null
07-18 19:23:29.034  5549  5570 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:29.034  5549  5570 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:29.037  5549  5568 D Listener: authHeader is null
07-18 19:23:29.037  5549  5568 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:23:29.037  5549  5568 V Listener: remoteURL() addr=127.0.0.1
07-18 19:23:29.051  5549  5571 I Database: Invalid filename as a view store: filename=db.forest.0
07-18 19:23:29.051  5549  5571 I Database: Invalid filename as a view store: filename=db.forest.meta
07-18 19:23:29.051  5549  5571 I Database: Invalid filename as a view store: filename=attachments
07-18 19:23:29.051  5549  5571 I Database: Invalid filename as a view store: filename=test%2Fbasic.viewindex.meta
07-18 19:23:29.051  5549  5571 I Database: Invalid filename as a view store: filename=test%2Fplayer.viewindex.meta
07-18 19:23:29.051  5549  5568 I Database: Invalid filename as a view store: filename=db.forest.0
07-18 19:23:29.051  5549  5568 I Database: Invalid filename as a view store: filename=db.forest.meta
07-18 19:23:29.051  5549  5568 I Database: Invalid filename as a view store: filename=attachments
07-18 19:23:29.051  5549  5568 I Database: Invalid filename as a view store: filename=test%2Fbasic.viewindex.meta
07-18 19:23:29.051  5549  5568 I Database: Invalid filename as a view store: filename=test%2Fplayer.viewindex.meta
07-18 19:23:29.052  5549  5571 I cbforest: Database: beginTransaction
07-18 19:23:29.052  5549  5571 I cbforest: Database: beginTransaction
07-18 19:23:29.053  5549  5570 I Database: Invalid filename as a view store: filename=db.forest.0
07-18 19:23:29.053  5549  5570 I Database: Invalid filename as a view store: filename=db.forest.meta
07-18 19:23:29.053  5549  5570 I Database: Invalid filename as a view store: filename=attachments
07-18 19:23:29.053  5549  5570 I Database: Invalid filename as a view store: filename=test%2Fbasic.viewindex.meta
07-18 19:23:29.053  5549  5570 I Database: Invalid filename as a view store: filename=test%2Fplayer.viewindex.meta
07-18 19:23:29.055  5549  5569 I Database: Invalid filename as a view store: filename=db.forest.0
07-18 19:23:29.055  5549  5569 I Database: Invalid filename as a view store: filename=db.forest.meta
07-18 19:23:29.055  5549  5569 I Database: Invalid filename as a view store: filename=attachments
07-18 19:23:29.055  5549  5569 I Database: Invalid filename as a view store: filename=test%2Fbasic.viewindex.meta
07-18 19:23:29.055  5549  5569 I Database: Invalid filename as a view store: filename=test%2Fplayer.viewindex.meta
07-18 19:27:45.809  1529  1542 I UsageStatsService: User[0] Flushing usage stats to disk

```

Happens with ForestDB as well
```
9  2712 I Database: Invalid filename as a view store: filename=db.forest.0
07-18 19:55:11.900  2689  2712 I Database: Invalid filename as a view store: filename=db.forest.meta
07-18 19:55:11.901  2689  2712 I Database: Invalid filename as a view store: filename=attachments
07-18 19:55:11.872  2689  2710 I Database: Invalid filename as a view store: filename=db.forest.0
07-18 19:55:11.912  2689  2710 I Database: Invalid filename as a view store: filename=db.forest.meta
07-18 19:55:11.913  2689  2710 I Database: Invalid filename as a view store: filename=attachments
07-18 19:55:11.914  2689  2710 I Database: Invalid filename as a view store: filename=test%2Fbasic.viewindex.meta
07-18 19:55:11.924  2689  2712 I Database: Invalid filename as a view store: filename=test%2Fbasic.viewindex.meta
07-18 19:55:11.882  2689  2711 I cbforest: Database: beginTransaction
07-18 19:55:11.925  2689  2710 E cbforest: Assertion failed: !_inTransaction (/home/couchbase/jenkins/workspace/couchbase-lite-android-build/couchbase-lite-android/libraries/couchbase-lite-java-forestdb/jni/../vendor/cbforest/CBForest/Database.cc:274, in void cbforest::Database::beginTransaction(cbforest::Transaction *))
07-18 19:55:11.932  2689  2712 E cbforest: Assertion failed: !_inTransaction (/home/couchbase/jenkins/workspace/couchbase-lite-android-build/couchbase-lite-android/libraries/couchbase-lite-java-forestdb/jni/../vendor/cbforest/CBForest/Database.cc:274, in void cbforest::Database::beginTransaction(cbforest::Transaction *))
07-18 19:55:11.947  2689  2711 I cbforest: Database: abort transaction
07-18 19:55:11.950  2689  2711 V View    : Query test/basic: Found row with key=cbl_views_5, value=bar, id=cbl_views_5
07-18 19:55:11.951  2689  2711 V View    : Query test/basic: Found row with key=cbl_views_6, value=bar, id=cbl_views_6
07-18 19:55:11.960  2689  2708 D Listener: authHeader is null
07-18 19:55:11.961  2689  2708 V Listener: Not enforcing basic auth -- allowedCredentials null or empty
07-18 19:55:11.981  2689  2710 W Router  : Router unable to route request to do_GET_DesignDocument
07-18 19:55:11.981  2689  2710 W Router  : java.lang.reflect.InvocationTargetException
07-18 19:55:11.981  2689  2710 W Router  : 	at java.lang.reflect.Method.invoke(Native Method)
07-18 19:55:11.981  2689  2710 W Router  : 	at com.couchbase.lite.router.Router.start(Router.java:571)
07-18 19:55:11.981  2689  2710 W Router  : 	at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
07-18 19:55:11.981  2689  2710 W Router  : 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
07-18 19:55:11.981  2689  2710 W Router  : 	at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2251)
07-18 19:55:11.981  2689  2710 W Router  : 	at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2176)
07-18 19:55:11.981  2689  2710 W Router  : 	at Acme.Serve.Serve$ServeConnection.run(Serve.java:1988)
07-18 19:55:11.981  2689  2710 W Router  : 	at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1223)
07-18 19:55:11.981  2689  2710 W Router  : 	at java.lang.Thread.run(Thread.java:818)
07-18 19:55:11.981  2689  2710 W Router  : Caused by: com.couchbase.lite.CouchbaseLiteException, Status: 590 (HTTP 500 Database error!)
07-18 19:55:11.981  2689  2710 W Router  : 	at com.couchbase.lite.store.ForestDBViewStore.updateIndexes(ForestDBViewStore.java:303)
07-18 19:55:11.981  2689  2710 W Router  : 	at com.couchbase.lite.View.updateIndexes(View.java:321)
07-18 19:55:11.981  2689  2710 W Router  : 	at com.couchbase.lite.View.updateIndex(View.java:294)
07-18 19:55:11.981  2689  2710 W Router  : 	at com.couchbase.lite.router.Router.queryDesignDoc(Router.java:2403)
07-18 19:55:11.981  2689  2710 W Router  : 	at com.couchbase.lite.router.Router.do_GET_DesignDocument(Router.java:2435)
07-18 19:55:11.981  2689  2710 W Router  : 	... 9 more
07-18 19:55:11.983  2689  2712 W Router  : Router unable to route request to do_GET_DesignDocument
07-18 19:55:11.983  2689  2712 W Router  : java.lang.reflect.InvocationTargetException
07-18 19:55:11.983  2689  2712 W Router  : 	at java.lang.reflect.Method.invoke(Native Method)
07-18 19:55:11.983  2689  2712 W Router  : 	at com.couchbase.lite.router.Router.start(Router.java:571)
07-18 19:55:11.983  2689  2712 W Router  : 	at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
07-18 19:55:11.983  2689  2712 W Router  : 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
07-18 19:55:11.983  2689  2712 W Router  : 	at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2251)
07-18 19:55:11.983  2689  2712 W Router  : 	at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2176)
07-18 19:55:11.983  2689  2712 W Router  : 	at Acme.Serve.Serve$ServeConnection.run(Serve.java:1988)
07-18 19:55:11.983  2689  2712 W Router  : 	at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1223)
07-18 19:55:11.983  2689  2712 W Router  : 	at java.lang.Thread.run(Thread.java:818)
07-18 19:55:11.983  2689  2712 W Router  : Caused by: com.couchbase.lite.CouchbaseLiteException, Status: 590 (HTTP 500 Database error!)
07-18 19:55:11.983  2689  2712 W Router  : 	at com.couchbase.lite.store.ForestDBViewStore.updateIndexes(ForestDBViewStore.java:303)
07-18 19:55:11.983  2689  2712 W Router  : 	at com.couchbase.lite.View.updateIndexes(View.java:321)
07-18 19:55:11.983  2689  2712 W Router  : 	at com.couchbase.lite.View.updateIndex(View.java:294)
07-18 19:55:11.983  2689  2712 W Router  : 	at com.couchbase.lite.router.Router.queryDesignDoc(Router.java:2403)
07-18 19:55:11.983  2689  2712 W Router  : 	at com.couchbase.lite.router.Router.do_GET_DesignDocument(Router.java:2435)
07-18 19:55:11.983  2689  2712 W Router  : 	... 9 more
07-18 19:55:11.984  2689  2712 V Listener: RouterCallbackBlock.onResponseReady() START
07-18 19:55:11.985  2689  2712 V Listener: RouterCallbackBlock.onResponseReady() END
07-18 19:55:11.985  2689  2712 V Listener: CountDownLatch.await() START
07-18 19:55:11.985  2689  2712 V Listener: CountDownLatch.await() END
07-18 19:55:11.988  2689  2708 V Listener: remoteURL() addr=127.0.0.1
07-18 19:55:11.952  2689  2711 V View    : Query test/basic: Found row with key=cbl_views_7, value=bar, id=cbl_views_7
07-18 19:55:12.015  2689  2711 V View    : Query test/basic: Found row with key=cbl_views_8, value=bar, id=cbl_views_8
07-18 19:55:12.016  2689  2711 V View    : Query test/basic: Found row with key=cbl_views_9, value=bar, id=cbl_views_9
07-18 19:55:12.018  2689  2711 V Listener: RouterCallbackBlock.onResponseReady() START
```

--- a/src/main/java/com/couchbase/lite/Database.java
+++ b/src/main/java/com/couchbase/lite/Database.java
@@ -126,6 +126,7 @@ public class Database implements StoreDelegate {
     private final Object lockPostingChangeNotifications = new Object();
     private final long startTime;
     private Timer purgeTimer;
+    private final Object lockViews = new Object();
 
     /**
      * Each database can have an associated PersistentCookieStore,
@@ -435,15 +436,17 @@ public class Database implements StoreDelegate {
      */
     @InterfaceAudience.Public
     public View getExistingView(String name) {
-        View view = views != null ? views.get(name) : null;
-        if (view != null)
-            return view;
+        synchronized (lockViews) {
+            View view = views != null ? views.get(name) : null;
+            if (view != null)
+                return view;
 
-        try {
-            return registerView(new View(this, name, false));
-        } catch (CouchbaseLiteException e) {
-            // View is not exist.
-            return null;
+            try {
+                return registerView(new View(this, name, false));
+            } catch (CouchbaseLiteException e) {
+                // View is not exist.
+                return null;
+            }
         }
     }
 
@@ -500,18 +503,20 @@ public class Database implements StoreDelegate {
      */
     @InterfaceAudience.Public
     public View getView(String name) {
-        View view = null;
-        if (views != null) {
-            view = views.get(name);
-        }
-        if (view != null) {
-            return view;
-        }
-        try {
-            return registerView(new View(this, name, true));
-        } catch (CouchbaseLiteException e) {
-            Log.w(TAG, "Error in registerView: error=" + e.getLocalizedMessage(), e);
-            return null;
+        synchronized (lockViews) {
+            View view = null;
+            if (views != null) {
+                view = views.get(name);
+            }
+            if (view != null) {
+                return view;
+            }
+            try {
+                return registerView(new View(this, name, true));
+            } catch (CouchbaseLiteException e) {
+                Log.w(TAG, "Error in registerView: error=" + e.getLocalizedMessage(), e);
+                return null;
+            }
         }
     }
 
@@ -688,7 +693,7 @@ public class Database implements StoreDelegate {
      */
     @InterfaceAudience.Public
     public String toString() {
-        return this.getClass().getName() + '[' + path + ']';
+        return this.getClass().getName() + "@" + Integer.toHexString(hashCode()) + '[' + path + ']';
     }
 
     ///////////////////////////////////////////////////////////////////////////

--- a/src/main/java/com/couchbase/lite/router/Router.java
+++ b/src/main/java/com/couchbase/lite/router/Router.java
@@ -2390,47 +2390,49 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
             }
         }
 
+        long lastSequenceIndexed;
+
         // according to functional test, view updateIndex() and query should be in sequence.
         synchronized (view) {
             // updateIndex() uses transaction internally, not necessary to apply syncrhonized.
             view.updateIndex();
-
-            long lastSequenceIndexed = view.getLastSequenceIndexed();
-
-            QueryOptions options = new QueryOptions();
-            //if the view contains a reduce block, it should default to reduce=true
-            if (view.getReduce() != null)
-                options.setReduce(true);
-            if (!getQueryOptions(options))
-                return new Status(Status.BAD_REQUEST);
-            if (keys != null)
-                options.setKeys(keys);
-
-            // Check for conditional GET and set response Etag header:
-            if (keys == null) {
-                long eTag = options.isIncludeDocs() ? db.getLastSequenceNumber() : lastSequenceIndexed;
-                if (cacheWithEtag(String.format(Locale.ENGLISH, "%d", eTag))) {
-                    return new Status(Status.NOT_MODIFIED);
-                }
-            }
-
-            // convert from QueryRow -> Map
-            List<QueryRow> queryRows = view.query(options);
-            List<Map<String, Object>> rows = new ArrayList<Map<String, Object>>();
-            for (QueryRow queryRow : queryRows) {
-                rows.add(queryRow.asJSONDictionary());
-            }
-
-            Map<String, Object> responseBody = new HashMap<String, Object>();
-            responseBody.put("rows", rows);
-            responseBody.put("total_rows", view.getCurrentTotalRows());
-            responseBody.put("offset", options.getSkip());
-            if (options.isUpdateSeq()) {
-                responseBody.put("update_seq", lastSequenceIndexed);
-            }
-            connection.setResponseBody(new Body(responseBody));
-            return new Status(Status.OK);
+            lastSequenceIndexed = view.getLastSequenceIndexed();
         }
+
+        QueryOptions options = new QueryOptions();
+        //if the view contains a reduce block, it should default to reduce=true
+        if (view.getReduce() != null)
+            options.setReduce(true);
+        if (!getQueryOptions(options))
+            return new Status(Status.BAD_REQUEST);
+        if (keys != null)
+            options.setKeys(keys);
+
+        // Check for conditional GET and set response Etag header:
+        if (keys == null) {
+            long eTag = options.isIncludeDocs() ? db.getLastSequenceNumber() : lastSequenceIndexed;
+            if (cacheWithEtag(String.format(Locale.ENGLISH, "%d", eTag))) {
+                return new Status(Status.NOT_MODIFIED);
+            }
+        }
+
+        // convert from QueryRow -> Map
+        List<QueryRow> queryRows = view.query(options);
+        List<Map<String, Object>> rows = new ArrayList<Map<String, Object>>();
+        for (QueryRow queryRow : queryRows) {
+            rows.add(queryRow.asJSONDictionary());
+        }
+
+        Map<String, Object> responseBody = new HashMap<String, Object>();
+        responseBody.put("rows", rows);
+        responseBody.put("total_rows", view.getCurrentTotalRows());
+        responseBody.put("offset", options.getSkip());
+        if (options.isUpdateSeq()) {
+            responseBody.put("update_seq", lastSequenceIndexed);
+        }
+        connection.setResponseBody(new Body(responseBody));
+        return new Status(Status.OK);
+
     }
 
     public Status do_GET_DesignDocument(Database _db, String designDocID, String viewName)

--- a/src/main/java/com/couchbase/lite/router/Router.java
+++ b/src/main/java/com/couchbase/lite/router/Router.java
@@ -2363,71 +2363,74 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
 
     private Status queryDesignDoc(String designDoc, String viewName, List<Object> keys)
             throws CouchbaseLiteException {
-        String tdViewName = String.format(Locale.ENGLISH, "%s/%s", designDoc, viewName);
-        // getExistingView is not thread-safe, but not access to db. In the database, it should protect instance variable
-        View view = db.getExistingView(tdViewName);
-        if (view == null || view.getMap() == null) {
-            // No TouchDB view is defined, or it hasn't had a map block assigned;
-            // see if there's a CouchDB view definition we can compile:
-            RevisionInternal rev = db.getDocument(String.format(Locale.ENGLISH, "_design/%s", designDoc), null, true);
-            if (rev == null) {
-                return new Status(Status.NOT_FOUND);
-            }
-            Map<String, Object> views = (Map<String, Object>) rev.getProperties().get("views");
-            Map<String, Object> viewProps = (Map<String, Object>) views.get(viewName);
-            if (viewProps == null) {
-                return new Status(Status.NOT_FOUND);
-            }
-            // If there is a CouchDB view, see if it can be compiled from source:
-            view = compileView(tdViewName, viewProps);
-            if (view == null) {
-                return new Status(Status.INTERNAL_SERVER_ERROR);
-            }
-        }
 
-        QueryOptions options = new QueryOptions();
+        View view;
 
-        //if the view contains a reduce block, it should default to reduce=true
-        if (view.getReduce() != null) {
-            options.setReduce(true);
-        }
-
-        if (!getQueryOptions(options)) {
-            return new Status(Status.BAD_REQUEST);
-        }
-        if (keys != null) {
-            options.setKeys(keys);
-        }
-
-        // updateIndex() uses transaction internally, not necessary to apply syncrhonized.
-        view.updateIndex();
-
-        long lastSequenceIndexed = view.getLastSequenceIndexed();
-
-        // Check for conditional GET and set response Etag header:
-        if (keys == null) {
-            long eTag = options.isIncludeDocs() ? db.getLastSequenceNumber() : lastSequenceIndexed;
-            if (cacheWithEtag(String.format(Locale.ENGLISH, "%d", eTag))) {
-                return new Status(Status.NOT_MODIFIED);
+        // make sure only one view instance per same view.
+        synchronized (db) {
+            String tdViewName = String.format(Locale.ENGLISH, "%s/%s", designDoc, viewName);
+            view = db.getExistingView(tdViewName);
+            if (view == null || view.getMap() == null) {
+                // No TouchDB view is defined, or it hasn't had a map block assigned;
+                // see if there's a CouchDB view definition we can compile:
+                RevisionInternal rev = db.getDocument(String.format(Locale.ENGLISH, "_design/%s", designDoc), null, true);
+                if (rev == null) {
+                    return new Status(Status.NOT_FOUND);
+                }
+                Map<String, Object> views = (Map<String, Object>) rev.getProperties().get("views");
+                Map<String, Object> viewProps = (Map<String, Object>) views.get(viewName);
+                if (viewProps == null) {
+                    return new Status(Status.NOT_FOUND);
+                }
+                // If there is a CouchDB view, see if it can be compiled from source:
+                view = compileView(tdViewName, viewProps);
+                if (view == null) {
+                    return new Status(Status.INTERNAL_SERVER_ERROR);
+                }
             }
         }
 
-        // convert from QueryRow -> Map
-        List<QueryRow> queryRows = view.query(options);
-        List<Map<String, Object>> rows = new ArrayList<Map<String, Object>>();
-        for (QueryRow queryRow : queryRows) {
-            rows.add(queryRow.asJSONDictionary());
-        }
+        // according to functional test, view updateIndex() and query should be in sequence.
+        synchronized (view) {
+            // updateIndex() uses transaction internally, not necessary to apply syncrhonized.
+            view.updateIndex();
 
-        Map<String, Object> responseBody = new HashMap<String, Object>();
-        responseBody.put("rows", rows);
-        responseBody.put("total_rows", view.getCurrentTotalRows());
-        responseBody.put("offset", options.getSkip());
-        if (options.isUpdateSeq()) {
-            responseBody.put("update_seq", lastSequenceIndexed);
+            long lastSequenceIndexed = view.getLastSequenceIndexed();
+
+            QueryOptions options = new QueryOptions();
+            //if the view contains a reduce block, it should default to reduce=true
+            if (view.getReduce() != null)
+                options.setReduce(true);
+            if (!getQueryOptions(options))
+                return new Status(Status.BAD_REQUEST);
+            if (keys != null)
+                options.setKeys(keys);
+
+            // Check for conditional GET and set response Etag header:
+            if (keys == null) {
+                long eTag = options.isIncludeDocs() ? db.getLastSequenceNumber() : lastSequenceIndexed;
+                if (cacheWithEtag(String.format(Locale.ENGLISH, "%d", eTag))) {
+                    return new Status(Status.NOT_MODIFIED);
+                }
+            }
+
+            // convert from QueryRow -> Map
+            List<QueryRow> queryRows = view.query(options);
+            List<Map<String, Object>> rows = new ArrayList<Map<String, Object>>();
+            for (QueryRow queryRow : queryRows) {
+                rows.add(queryRow.asJSONDictionary());
+            }
+
+            Map<String, Object> responseBody = new HashMap<String, Object>();
+            responseBody.put("rows", rows);
+            responseBody.put("total_rows", view.getCurrentTotalRows());
+            responseBody.put("offset", options.getSkip());
+            if (options.isUpdateSeq()) {
+                responseBody.put("update_seq", lastSequenceIndexed);
+            }
+            connection.setResponseBody(new Body(responseBody));
+            return new Status(Status.OK);
         }
-        connection.setResponseBody(new Body(responseBody));
-        return new Status(Status.OK);
     }
 
     public Status do_GET_DesignDocument(Database _db, String designDocID, String viewName)

